toolchain/lower/handle.cpp

```diff
@@ -260,15 +262,23 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
       return;
     }
     case SemIR::BuiltinFunctionKind::IntDiv: {
-      context.SetLocal(inst_id, context.builder().CreateSDiv(
-                                    context.GetValue(arg_ids[0]),
-                                    context.GetValue(arg_ids[1]), "div"));
+      context.SetLocal(inst_id, IsSignedInt(context, inst_id)
```

Comment:
FWIW, I'm a little surprised you don't either make Uint* for all the operations, or more generally do a BuiltinDiv that accepts whatever types (since it could morph based on arguments, as here)

Reply:
There are definitely some other options we could pursue here. I'm going to land this as-is because I don't think any of them is obviously substantially better than the rest (and this patch is just fixing a bug here; most of the builtins already dispatched on signedness), but I think we should keep discussing this especially as we add the bigint type and floating-point support.

Some thoughts on specific approaches:

- Current approach: it's a bit weird that we have builtins with different semantics (eg, overflow behavior) with the same name.
- Provide an `Int*` set and a `UInt*` set of builtins: we'd end up with some redundancy for the operations that are identical for signed and unsigned types (eg, bitwise operations).
- Provide a more general `BuiltinDiv` that also supports floating-point types too: the implementation of that (especially in `eval.cpp`) would be substantially different in that case, because the types have a different representation, whereas it's pretty easy to handle all the integer types together, with just a small branch for signed vs unsigned operations in the cases where it matters.
- Provide different signed builtins for operations where it matters, eg add `int.sdiv` and `int.udiv` but keep a single `int.and`, and leave it up to the library to map the operations to the right type based on signedness: maybe this would be a nice option to pursue, but would add a bunch of builtin names. Would also support providing (eg) a wrapping add on signed integers as a library function, if desired. Matches LLVM better.

The last of those options seems pretty tempting to me.

Reply:
+1 to the last of those options from me FWIW -- mirroring the LLVM model for the builtins makes a lot of sense to me.